### PR TITLE
Update Python 3 compatibility. WIP.

### PR DESCRIPTION
**Pull request by janwijbrand**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
-develop-eggs/
-parts/
+__pycache__/
+.coverage
 .installed.cfg
+.mr.developer.cfg
+.tox/
+.tox/
+*.egg-info/
+*.egg-info/
```
Comment (mgedmin) [inline .gitignore]: `.tox/` and `*.egg-info/` appear twice?

File: .gitignore [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
-__pycache__/
+bin/
 build/
+develop-eggs/
 dist/
-*.egg-info/
-.tox/
-.coverage
 htmlcov/
+parts
```
Comment (mgedmin) [inline .gitignore]: Why drop the `/` from `parts/`?

File: .travis.yml [hunk 3]
```diff
@@ -2,14 +2,15 @@ language: python
...
     - 3.4
     - 3.5
     - 3.6
+    - pypy
+    - pypy3
 install:
     - pip install -U pip setuptools
     - pip install -U coverage coveralls
-    - pip install -U -e .[test]
+    - pip install -U -e .[test,browser]
```
Comment (mgedmin) [inline .travis.yml]: Perhaps `[test]` should include everything in `[browser]`?

File: buildout.cfg [hunk 4]
```diff
@@ -31,3 +36,6 @@ domain = zope.errorview
 output = src/zope/errorview/i18n/locales
 zcml =
   <include package="zope.i18n" file="meta.zcml" />
+
+[sources]
+z3c.recipe.i18n = git ${buildout:github}/z3c.recipe.i18n pushurl=${buildout:github_push}/z3c.recipe.i18n branch=py3
```
Comment (mgedmin) [inline buildout.cfg]: You plan to remove these once a Python-3-capable z3c.recipe.i18n is released to PyPI?

File: src/zope/errorview/browser.py [hunk 5]
```diff
@@ -52,6 +52,8 @@ def update(self):
             'Cache-Control', 'no-store, no-cache, must-revalidate')
         self.request.response.setHeader(
             'Pragma', 'no-cache')
+        self.request.response.setHeader(
+            'Content-Type', 'text/plain;charset=utf-8')
```
Comment (mgedmin) [inline src/zope/errorview/browser.py:56]: This doesn't seem to be related to Python 3 changes?

File: src/zope/errorview/http.py [hunk 6]
```diff
@@ -40,6 +40,8 @@ def __init__(self, context, request):
 
     def update(self):
         self.request.response.setStatus(500)
+        self.request.response.setHeader(
+            'Content-Type', 'text/plain; charset=utf-8')
```
Comment (mgedmin) [inline src/zope/errorview/http.py:44]: And this.

**Review by mgedmin — COMMENTED** [review]
Mostly LGTM.

Missing bits:

- setup.py updates
- changelog

Things I'm worried about:

- Content-Type change to text/plain. If people are subclassing zope.errorpage views and rendering HTML, their sites will break.  This may be fine, if we bump the major version number and put a prominent notice in the changelog.

File: .travis.yml (hunk 3 above)
Comment (mgedmin) [reply .travis.yml]: In fact setup.py already makes sure that `[test]` is a superset of `[browser]`.

File: src/zope/errorview/browser.py (hunk 5 above)
Comment (janwijbrand) [reply src/zope/errorview/browser.py:56]: This relates to one of your general remarks above.

These changes are in fact related to Python 3. Since in Python 2 when returning "just" a string as a response would have `zope.publisher` not require a content-type to be set as the response body would not be unicode.

In Python 3 "just" returning a string as response obviously *is* unicode and `zope.publisher` then requires (otherwise https://github.com/zopefoundation/zope.publisher/blob/c1ba72f6a3838269badc433ba9f06a4f1a0eabf2/src/zope/publisher/http.py#L821 would fail) a content type to be set. In other words, it became more important than before to make sure you set the correct content type in your application for your views - which in itself is a good thing I think.

I'll make a note of this in CHANGES.rst indeed and bump the version to 1.1.

File: src/zope/errorview/http.py (hunk 6 above)
Comment (janwijbrand) [reply src/zope/errorview/http.py:44]: Idem.

**Comment by janwijbrand** [comment]
@mgedmin thanks for the review. I think I processed all comments. If you're satisfied I'll make a release (unblocking further work on groktoolkit).

File: src/zope/errorview/browser.py (hunk 5 above)
Comment (mgedmin) [reply src/zope/errorview/browser.py:56]: Thank you for explaining!

File: buildout.cfg [hunk 7]
```diff
@@ -7,9 +7,9 @@ parts =
   i18n
 github = git://github.com/zopefoundation
 github_push = git@github.com:zopefoundation
-extensions = mr.developer
-auto-checkout =
-  z3c.recipe.i18n
+
+[versions]
+z3c.recipe.i18n = 1.0.0
```
Comment (mgedmin) [inline buildout.cfg]: I'm not too keen on pinning exact versions: we want to know if newer z3c.recipe.i18n releases somehow break us.

If you need to specify a minimum, buildout lets you say `zc.recipe.i18n = >= 1.0.0`.

**Review by mgedmin — APPROVED** [review]
LGTM with one caveat, but I don't think that's a blocker.

File: buildout.cfg (hunk 7 above)
Comment (janwijbrand) [reply buildout.cfg]: will do.

File: src/zope/errorview/http.py (hunk 6 above)
Comment (gyst) [reply src/zope/errorview/http.py:44]: This has a spurious extra space.

Comment (mgedmin) [reply src/zope/errorview/http.py:44]: Which space is spurious?

Comment (gyst) [reply src/zope/errorview/http.py:44]: Seems curious that the request header in `browser` should *not* have a space after the semicolon, while the response header here does add that space.
